pkg/bicopy/bicopy.go

```diff
@@ -26,56 +7,25 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// Bicopy is from https://github.com/rootless-containers/rootlesskit/blob/v0.10.1/pkg/port/builtin/parent/tcp/tcp.go#L73-L104
-// (originally from libnetwork, Apache License 2.0).
-func Bicopy(x, y io.ReadWriter, quit <-chan struct{}) {
-	type closeReader interface {
-		CloseRead() error
-	}
-	type closeWriter interface {
-		CloseWrite() error
-	}
-	var wg sync.WaitGroup
-	broker := func(to, from io.ReadWriter) {
-		if _, err := io.Copy(to, from); err != nil {
-			logrus.WithError(err).Debug("failed to call io.Copy")
-		}
-		if fromCR, ok := from.(closeReader); ok {
-			if err := fromCR.CloseRead(); err != nil {
-				logrus.WithError(err).Debug("failed to call CloseRead")
-			}
-		}
-		if toCW, ok := to.(closeWriter); ok {
-			if err := toCW.CloseWrite(); err != nil {
-				logrus.WithError(err).Debug("failed to call CloseWrite")
-			}
-		}
-		wg.Done()
-	}
+type NamedReadWriter struct {
+	ReadWriter io.ReadWriter
+	Name       string
+}
 
+func Bicopy(context string, x, y NamedReadWriter) {
+	var wg sync.WaitGroup
 	wg.Add(2)
-	go broker(x, y)
-	go broker(y, x)
-	finish := make(chan struct{})
 	go func() {
-		wg.Wait()
-		close(finish)
-	}()
-
-	select {
-	case <-quit:
-	case <-finish:
-	}
-	if xCloser, ok := x.(io.Closer); ok {
-		if err := xCloser.Close(); err != nil {
-			logrus.WithError(err).Debug("failed to call xCloser.Close")
+		defer wg.Done()
+		if _, err := io.Copy(x.ReadWriter, y.ReadWriter); err != nil {
+			logrus.WithError(err).Errorf("%s: io.Copy(%s, %s)", context, x.Name, y.Name)
 		}
-	}
-	if yCloser, ok := y.(io.Closer); ok {
-		if err := yCloser.Close(); err != nil {
-			logrus.WithError(err).Debug("failed to call yCloser.Close")
+	}()
+	go func() {
+		defer wg.Done()
+		if _, err := io.Copy(y.ReadWriter, x.ReadWriter); err != nil {
+			logrus.WithError(err).Errorf("%s: io.Copy(%s, %s)", context, y.Name, x.Name)
 		}
-	}
-	<-finish
-	// TODO: return copied bytes
+	}()
+	wg.Wait()
```

Comment:
Do we really need bicopy to handle the error and not throw it to the caller ??

I would prefer it to be handled by the caller itself. It gives us better control. Handling error ourself looks generic but all we are doing is just logging. But in some cases we may need to act on error in caller side

Reply:
There has been a lot of back and forth on this already. I'm inclined to leave it as is. Another option is to pass a handler callback, but nobody needs this today.

Reply:
`but nobody needs this today` - If errors are returned back, callers will have more control to handle logging parts. Some cases might need more detailed logging rather than generic message `context: io.Copy(y, x)` which is not self explanatory.

If you have a valid reason for not returning error please let me know. 

Also i think `func Bicopy(context string, x, y NamedReadWriter) error {}` should be good as well. We don't need to complicate with handlers / callbacks

Reply:
@nirs expressed concern that if `io.Copy` in one direction returns, the `io.Copy` in the other direction might not - and returning an error cannot happen until both functions complete. That's why the latest versions logs directly instead of returning an error.

Reply:
It will eventually fail because one end of stream is closed already. 

Since we have wg.Done at last. Returning error after both io.Copy are closed looks correct to me.

Reply:
@nirs What do you think ??

Reply:
We could even use errgroup. Which will close the group when a non-nil error is returned. 

Code can look something like

```
wg, ctx := errgroup.WithContext(ctx)
defer b.Close()
defer a.Close()

wg.Go(func() error {
         _, err1 = io.Copy(a, b)
	return err1
})
wg.Go(func() error {
	_, err1 = io.Copy(b, a)
	return err1
})
return wg.Wait()
```

Even in case close() is not there, we will not do close but all other stays the same

Reply:
With errgroup as soon as one is returned non-nil its over. It will close the other as well.

If we are using reader/writer then we will not even have close so it's just stopping the loop

Reply:
The goroutine is canceled magically? how does it stop?

Reply:
I don't see anything in the docs about cancelling goroutine blocked on a syscall.
https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext

Reply:
Yes you are right, My mistake.

Only context is closed. So we need to check for context close and close the connection.

For pure reader / writer until that syscall is returned we have no other option. But most of our reader / writer have close method so io.Closable should be fine

Reply:
I think this should also be added to tcpproxy upstream - ensure termination and return an error. With both we can use it instead of Bicopy.